### PR TITLE
Optimizing the interface of RestServer and RestClient

### PR DESCRIPTION
**Pull request by Patrick0308**
**What this PR does**:
To define RestServerRequest and RestServerResponse, make it very easy for users to implement RestServer interface.

### CODE REVIEW

File: protocol/rest/server/rest_server.go [hunk 1]
```diff
@@ -17,15 +17,274 @@
...
+)
+
 import (
 	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/logger"
 	"github.com/apache/dubbo-go/protocol"
-	"github.com/apache/dubbo-go/protocol/rest/config"
+	"github.com/apache/dubbo-go/protocol/invocation"
+	rest_config "github.com/apache/dubbo-go/protocol/rest/config"
 )
```
Comment (AlexStocks) [inline protocol/rest/server/rest_server.go:39]: add comment for this interface

File: protocol/rest/server/rest_server.go [hunk 2]
```diff
@@ -17,15 +17,274 @@
...
 
 type RestServer interface {
 	Start(url common.URL)
```
Comment (AlexStocks) [inline protocol/rest/server/rest_server.go:41]: add comment for interface funcs

File: protocol/rest/server/rest_server.go [hunk 3]
```diff
@@ -17,15 +17,274 @@
...
-	Deploy(invoker protocol.Invoker, restMethodConfig map[string]*config.RestMethodConfig)
-	UnDeploy(restMethodConfig map[string]*config.RestMethodConfig)
+	Deploy(restMethodConfig *rest_config.RestMethodConfig, routeFunc func(request RestServerRequest, response RestServerResponse))
+	UnDeploy(restMethodConfig *rest_config.RestMethodConfig)
 	Destroy()
 }
+
+// RestServerRequest interface
+type RestServerRequest interface {
+	RawRequest() *http.Request
```
Comment (AlexStocks) [inline protocol/rest/server/rest_server.go:49]: add comments for interface funcs

File: protocol/rest/server/rest_server.go [hunk 4]
```diff
@@ -17,15 +17,274 @@
...
+	QueryParameter(name string) string
+	QueryParameters(name string) []string
+	BodyParameter(name string) (string, error)
+	HeaderParameter(name string) string
+	ReadEntity(entityPointer interface{}) error
+}
+
+// RestServerResponse interface
+type RestServerResponse interface {
+	Header() http.Header
```
Comment (AlexStocks) [inline protocol/rest/server/rest_server.go:61]: add comments for interface funcs

File: protocol/rest/server/rest_server.go [hunk 5]
```diff
@@ -17,15 +17,274 @@
...
+	Write([]byte) (int, error)
+	WriteHeader(statusCode int)
+	WriteError(httpStatus int, err error) (writeErr error)
+	WriteEntity(value interface{}) error
+}
+
+func GetRouteFunc(invoker protocol.Invoker, methodConfig *rest_config.RestMethodConfig) func(req RestServerRequest, resp RestServerResponse) {
```
Comment (AlexStocks) [inline protocol/rest/server/rest_server.go:68]: add comment

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 6]
```diff
@@ -80,58 +76,25 @@ func (grs *GoRestfulServer) Start(url common.URL) {
...
-		method := svc.Method()[methodName]
-		argsTypes := method.ArgsType()
-		replyType := method.ReplyType()
-		ws := new(restful.WebService)
-		ws.Path(config.Path).
-			Produces(strings.Split(config.Produces, ",")...).
-			Consumes(strings.Split(config.Consumes, ",")...).
-			Route(ws.Method(config.MethodType).To(getFunc(methodName, invoker, argsTypes, replyType, config)))
-		grs.container.Add(ws)
+func (grs *GoRestfulServer) Deploy(restMethodConfig *config.RestMethodConfig, routeFunc func(request server.RestServerRequest, response server.RestServerResponse)) {
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:79]: add comment

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 7]
```diff
@@ -80,58 +76,25 @@ func (grs *GoRestfulServer) Start(url common.URL) {
...
-	}
-}
-func (grs *GoRestfulServer) UnDeploy(restMethodConfig map[string]*config.RestMethodConfig) {
-	for _, config := range restMethodConfig {
-		ws := new(restful.WebService)
-		ws.Path(config.Path)
-		err := grs.container.Remove(ws)
-		if err != nil {
-			logger.Warnf("[Go restful] Remove web service error:%v", err)
-		}
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:134]: add comment

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 8]
```diff
@@ -80,58 +76,25 @@ func (grs *GoRestfulServer) Start(url common.URL) {
...
+func (grs *GoRestfulServer) UnDeploy(restMethodConfig *config.RestMethodConfig) {
+	ws := new(restful.WebService)
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:93]: ws := &restful.WebService{}. Pls do not use new.

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 9]
```diff
@@ -144,179 +107,54 @@ func (grs *GoRestfulServer) Destroy() {
...
-		if t.Kind() == reflect.Ptr {
-			t = t.Elem()
-		}
-		if t.Kind() == reflect.String {
-			args[k] = param
-		} else {
-			logger.Errorf("[Go restful] Header param parse error, the args:%v of type isn't string", k)
-		}
-	}
+func NewGoRestfulRequestAdapter(request *restful.Request) *GoRestfulRequestAdapter {
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:126]: add comment

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 10]
```diff
@@ -144,179 +107,54 @@ func (grs *GoRestfulServer) Destroy() {
...
+	return &GoRestfulRequestAdapter{request: request}
+}
 
-	return args
+func (grra *GoRestfulRequestAdapter) RawRequest() *http.Request {
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:130]: add comment a

File: protocol/rest/server/server_impl/go_restful_server.go [hunk 11]
```diff
@@ -144,179 +107,54 @@ func (grs *GoRestfulServer) Destroy() {
...
+	return grra.request.Request
 }
 
-func GetNewGoRestfulServer() server.RestServer {
-	return NewGoRestfulServer()
+func (grra *GoRestfulRequestAdapter) PathParameter(name string) string {
```
Comment (AlexStocks) [inline protocol/rest/server/server_impl/go_restful_server.go:134]: add comment

File: protocol/rest/server/server_impl/go_restful_server.go (hunk 8 above)
Comment (Patrick0308) [reply protocol/rest/server/server_impl/go_restful_server.go:100]: Why do not use the function of new?

Comment (AlexStocks) [reply protocol/rest/server/server_impl/go_restful_server.go:100]: > Why do not use the function of new?

as u know, when new a java object, its actual action is as follows,
1 allocate a memory space for the obj;
2 invoke its construct to init the obj;

however, the `new` just allocate a memory space for the obj in Go.

File: protocol/rest/server/rest_server.go [hunk 12]
```diff
@@ -17,15 +17,288 @@
...
+		var (
+			err  error
+			args []interface{}
+		)
+		svc := common.ServiceMap.GetService(invoker.GetUrl().Protocol, strings.TrimPrefix(invoker.GetUrl().Path, "/"))
+		// get method
+		method := svc.Method()[methodConfig.MethodName]
+		argsTypes := method.ArgsType()
+		replyType := method.ReplyType()
+		if (len(argsTypes) == 1 || len(argsTypes) == 2 && replyType == nil) &&
```
Comment (flycash) [inline protocol/rest/server/rest_server.go:93]: Please add some comments here. It's a little bit hard to understand and I'm afraid many users cannot understand why we write those if-else statements.

File: protocol/rest/server/rest_server.go [hunk 13]
```diff
@@ -17,15 +17,288 @@
...
+			i64, err = strconv.ParseInt(req.PathParameter(v), 10, 32)
+			if err == nil {
+				param = int32(i64)
+			}
+		} else if kind == reflect.Int64 {
+			param, err = strconv.ParseInt(req.PathParameter(v), 10, 64)
+		} else if kind == reflect.String {
+			param = req.PathParameter(v)
+		} else {
+			logger.Warnf("[Go restful] Path param parse error, the args:%v of type isn't int or string", k)
```
Comment (flycash) [inline protocol/rest/server/rest_server.go:295]: Using Errorf to be aligned with `assembleArgsFromQueryParams`. I have a question, should we continue here? I think if parsing the param got  error, there must be some problem in user's codes. Just return would be better.
It's just a advice.

**Comment by Patrick0308** [comment]
I remove the category of consumer's url. It test ok.

**Pull request merged**